### PR TITLE
ivpn{,-service}: 3.10.0 -> 3.10.15, init ivpn module

### PR DESCRIPTION
**Pull request by AtaraxiaSjel**
###### Description of changes

This PR fixes ivpn-service static paths to binary and init ivpn module (automatically run ivpn-service).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AtaraxiaSjel** [comment]
Result of `nixpkgs-review pr 226514` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ivpn</li>
    <li>ivpn-service</li>
  </ul>
</details>

**Review by jopejoe1 — APPROVED** [review]
Tested the module, and it's working on my system.

File: pkgs/tools/networking/ivpn/default.nix [hunk 1]
```diff
@@ -31,16 +40,58 @@ builtins.mapAttrs (pname: attrs: buildGoModule (attrs // rec {
...
     vendorHash = "sha256-9Rk6ruMpyWtQe+90kw4F8OLq7/JcDSrG6ufkfcrS4W8=";
-    buildInputs = [ wirelesstools ];
+    nativeBuildInputs = [ makeWrapper ];
+    buildInputs = [
+      wirelesstools
+      wireguard-tools
+      openvpn
+      obfs4
+      iproute2
+      dnscrypt-proxy2
```
Comment (urandom2) [inline pkgs/tools/networking/ivpn/default.nix:60]: can we sort these:
```suggestion
      dnscrypt-proxy2
      iproute2
      obfs4
      openvpn
      wireguard-tools
      wirelesstools
```

**Review by urandom2 — APPROVED** [review]
no comment on the nixos module, derivation looks good

**Comment by AtaraxiaSjel** [comment]
@urandom2 np, done!

File: pkgs/tools/networking/ivpn/default.nix [hunk 2]
```diff
@@ -31,16 +40,58 @@ builtins.mapAttrs (pname: attrs: buildGoModule (attrs // rec {
...
+
+    postFixup = ''
+      install -Dm700 $src/daemon/References/Linux/etc/client.down "$out/etc/client.down"
+      install -Dm700 $src/daemon/References/Linux/etc/client.up "$out/etc/client.up"
+      install -Dm700 $src/daemon/References/Linux/etc/firewall.sh "$out/etc/firewall.sh"
+      install -Dm700 $src/daemon/References/Linux/etc/splittun.sh "$out/etc/splittun.sh"
+      install -Dm600 $src/daemon/References/common/etc/servers.json "$out/etc/servers.json"
+      install -Dm400 $src/daemon/References/common/etc/ca.crt "$out/etc/ca.crt"
+      install -Dm400 $src/daemon/References/common/etc/ta.key "$out/etc/ta.key"
+      install -Dm400 $src/daemon/References/common/etc/dnscrypt-proxy-template.toml "$out/etc/dnscrypt-proxy-template.toml"
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/ivpn/default.nix:77]: We can't just wildcard copy the files?

File: pkgs/tools/networking/ivpn/default.nix [hunk 3]
```diff
@@ -31,16 +40,58 @@ builtins.mapAttrs (pname: attrs: buildGoModule (attrs // rec {
...
     vendorHash = "sha256-9Rk6ruMpyWtQe+90kw4F8OLq7/JcDSrG6ufkfcrS4W8=";
-    buildInputs = [ wirelesstools ];
+    nativeBuildInputs = [ makeWrapper ];
+    buildInputs = [
+      dnscrypt-proxy2
+      iproute2
+      obfs4
+      openvpn
+      wireguard-tools
+      wirelesstools
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/ivpn/default.nix:60]: Are those even used somewhere? buildInputs is not affecting the usage of wrapProgram below.

File: pkgs/tools/networking/ivpn/binary-paths.patch [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
++			wgToolBinaryPath = p
++		}
++	}
+
+-	dnscryptproxyBinPath = path.Join(installDir, "dnscrypt-proxy/dnscrypt-proxy")
++	if p, err := exec.LookPath("dnscrypt-proxy"); err == nil {
++		if p, err = filepath.Abs(p); err == nil {
++			dnscryptproxyBinPath = p
++		}
++	}
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/ivpn/binary-paths.patch:122]: We could just hardcode those paths into the code with nix and drop the error checking

File: nixos/modules/services/networking/ivpn.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    networking.firewall.checkReversePath = "loose";
+
+    systemd.services.ivpn-service = {
+      description = "iVPN daemon";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network.target" ];
+      after = [
+        "network-online.target"
+        "NetworkManager.service"
+        "systemd-resolved.service"
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ivpn.nix:34]: Does this work if network manager or resolved are disabled?

File: pkgs/tools/networking/ivpn/default.nix (hunk 3 above)
Comment (AtaraxiaSjel) [reply pkgs/tools/networking/ivpn/default.nix]: It was needed before, but with the new patch (hardcoded paths to binaries), it is no longer necessary.

File: pkgs/tools/networking/ivpn/default.nix (hunk 2 above)
Comment (AtaraxiaSjel) [reply pkgs/tools/networking/ivpn/default.nix]: done

File: pkgs/tools/networking/ivpn/default.nix [hunk 6]
```diff
@@ -31,16 +40,58 @@ builtins.mapAttrs (pname: attrs: buildGoModule (attrs // rec {
...
+      install -Dm700 $src/daemon/References/Linux/etc/splittun.sh "$out/etc/splittun.sh"
+      install -Dm600 $src/daemon/References/common/etc/servers.json "$out/etc/servers.json"
+      install -Dm400 $src/daemon/References/common/etc/ca.crt "$out/etc/ca.crt"
+      install -Dm400 $src/daemon/References/common/etc/ta.key "$out/etc/ta.key"
+      install -Dm400 $src/daemon/References/common/etc/dnscrypt-proxy-template.toml "$out/etc/dnscrypt-proxy-template.toml"
+
+      patchShebangs --build $out/etc/firewall.sh
+      patchShebangs --build $out/etc/splittun.sh
+      patchShebangs --build $out/etc/client.down
+      patchShebangs --build $out/etc/client.up
```
[the remark this answers is not in the document]
Comment (AtaraxiaSjel) [reply pkgs/tools/networking/ivpn/default.nix]: done

File: pkgs/tools/networking/ivpn/binary-paths.patch (hunk 4 above)
Comment (AtaraxiaSjel) [reply pkgs/tools/networking/ivpn/binary-paths.patch]: i changed it to substituteinplace.

File: nixos/modules/services/networking/ivpn.nix (hunk 5 above)
Comment (AtaraxiaSjel) [reply nixos/modules/services/networking/ivpn.nix:34]: systemd-resolved doesn't run on my system, but everything works perfectly. I copied this from the mullvad-vpn module, but if it is necessary, I will check it later.

**Comment by SuperSandro2000** [comment]
@ofborg build ivpn ivpn-service

**Pull request merged**